spring-integration-mail/src/main/java/org/springframework/integration/mail/AbstractMailReceiver.java

```diff
@@ -503,17 +503,27 @@ private Object byteArrayToContent(Map<String, Object> headers, ByteArrayOutputSt
 	}
 
 	private void postProcessFilteredMessages(Message[] filteredMessages) throws MessagingException {
-		setMessageFlags(filteredMessages);
-
-		if (shouldDeleteMessages()) {
-			deleteMessages(filteredMessages);
-		}
 		// Copy messages to cause an eager fetch
 		if (this.headerMapper == null && (this.autoCloseFolder || this.simpleContent)) {
+			Message[] originalMessages = new Message[filteredMessages.length];
 			for (int i = 0; i < filteredMessages.length; i++) {
-				MimeMessage mimeMessage = new IntegrationMimeMessage((MimeMessage) filteredMessages[i]);
+				Message originalMessage = filteredMessages[i];
+				originalMessages[i] = originalMessage;
+				MimeMessage mimeMessage = new IntegrationMimeMessage((MimeMessage) originalMessage);
 				filteredMessages[i] = mimeMessage;
 			}
+			setMessageFlagsAndMaybeDeleteMessages(originalMessages);
+		}
+		else {
+			setMessageFlagsAndMaybeDeleteMessages(filteredMessages);
```

Comment:
May we consider a local method variable for `Message[]` instead of `else` branch with the same method call?

I mean something like:
```
Message[] messages = filteredMessages;
if (this.headerMapper == null && (this.autoCloseFolder || this.simpleContent)) {
    messages = new Message[filteredMessages.length];
    ...
}
setMessageFlagsAndMaybeDeleteMessages(messages);
```

Reply:
I was wanted to do that initially as well. However, in the `if` we are creating a new `Message` of the `IntegrationMimeMessage`. This means that when we call `setMessageFlagsAndMaybeDeleteMessages` and we invoke `Message#setFlags` the actual implementation differs. If we look into the `IMAPMessage#setFlags` the implementation looks like

```java
synchronized (getMessageCacheLock()) {
    try {
        IMAPProtocol p = getProtocol();
        checkExpunged(); // Insure that this message is not expunged
        p.storeFlags(getSequenceNumber(), flag, set);
    } catch (ConnectionException cex) {
        throw new FolderClosedException(folder, cex.getMessage());
    } catch (ProtocolException pex) {
        throw new MessagingException(pex.getMessage(), pex);
    }
}
```

whereas the `MimeMessage#setFlags` looks like

```java
if (set)
    flags.add(flag);
else
    flags.remove(flag);
```

I am going to add a comment there to explain why it is done in the way I did it

Reply:
I’m not sure why does that matter since you call `setMessageFlagsAndMaybeDeleteMessages` in both cases? So, one local variable for both cases feels right.
Ok, I’ll look into that locally on merge.

Reply:
@artembilan it matters a lot because what will be invoked in the end is not the same. The behaviour of the `setFlags` will change due to the fact that there will be different implementations. 

Prior to this PR the `setFlags` was invoked on the original messages received from the folder. The goal of the PR is to still invoke the `setFlags` on the original messages. However, instead of doing it before copying it, to do it after it has been copied.

I have an idea for a test case. I'll add it to illustrate the problem.